src/triage/component/postmodeling/plots.py

```diff
@@ -0,0 +1,211 @@
+
+def _store(fig, **kwargs):
+    output_type=kwargs.get('output_type', 'show')
+
+    if (output_type == 'save'):
+        filename = kwargs.get('filename', f"{model}")
+        extension = kwargs.get('extension', 'png')
+        transparent = kwargs.get('transparent', False)
+        dpi = kwargs.get('dpi', 80)
+        bbox_inches = kwargs.get('bbox_inches', 'tight')
+        fig.savefig(f"{filename}.{extension}", transparent=transparent, dpi=dpi, bbox_inches=bbox_inches)
+
+def plot_roc(model, **kwargs):
+    fig, ax = plt.subplots()
+
+    predictions = get_predictions(model)
+    labels = predictions.label_value
+    scores = predictions.score
+    fpr, tpr, thresholds = roc_curve(labels, scores)
+    roc_auc = auc(fpr, tpr)
+
+    fmt = kwargs.get('fmt', 'b')
+
+    # Random performance
+    ax.plot([0, 1], [0, 1], 'k--')
+
+    # Actual model
+    ax.plot(fpr, tpr, fmt, label='ROC curve (area = %0.2f)' % roc_auc)
+
+    ax.set_xlim([0.0, 1.0])
+    ax.set_ylim([0.0, 1.0])
+
+    ax.set_xlabel('False Positive Rate')
+    ax.set_ylabel('True Positive Rate')
+
+    ax.set_title(kwargs.get('title', f"(Model {model.id} @ {model.train_end_time.strftime('%Y-%m-%d')})"))
+
+    ax.legend(loc="lower right")
+
+    fig.suptitle(kwargs.get('suptitle', f"Model Group {model.model_group.id}"))
+
+    _store(fig, **kwargs)
+
+    return fig, ax
+
+
+def plot_precision_recall_n(model, **kwargs):
+    fig, ax1 = plt.subplots()
+
+    predictions = get_predictions(model)
+    labels = predictions.label_value
+    scores = predictions.score
+
+    y_axis = np.arange(0, 101, 1)
+
+    prec_k = precision_n(labels, scores)
+    rec_k = recall_n(labels, scores)
+
+    prec_color = kwargs.get('prec_fmt', 'tab:blue')
+    rec_color = kwargs.get('rec_fmt', 'tab:red')
+
+    ax1.plot(y_axis, prec_k, prec_color)
+    ax1.set_xlabel('Percent of population')
+    ax1.set_ylabel('Precision', color=prec_color)
+    ax1.tick_params(axis='y', labelcolor=prec_color)
+
+    ax1.set_ylim([0,1])
+    ax1.set_xlim([0,100])
+
+    ax1.set_title(kwargs.get('title', f"(Model {model.id} @ {model.train_end_time.strftime('%Y-%m-%d')})"))
+
+    ax2 = ax1.twinx()
+    ax2.plot(y_axis, rec_k, rec_color)
+    ax2.set_ylim([0,1])
+    ax2.set_ylabel('Recall', color=rec_color)
+    ax2.tick_params(axis='y', labelcolor=rec_color)
+
+
+    fig.suptitle(kwargs.get('suptitle', f"Model Group {model.model_group.id}"))
+
+    _store(fig, **kwargs)
+
+    return fig, ax1, ax2
+
+def plot_metric_over_time(model_groups, metric, parameter, **kwargs):
+
+    def get_model_group_evaluations(model_group):
+        evaluations = pd.concat([get_evaluations(model).query(f"metric == '{metric}@' and parameter == '{parameter}'") for model in model_group])
+        evaluations = evaluations.sort_values('evaluation_start_time', ascending=True)
+        evaluations = evaluations[['evaluation_start_time', 'value']]
+        evaluations['type'] = model_group.type.split('.')[-1]
+        return evaluations
```

Comment:
this function needn't be declared dynamically, need it?

perhaps it'll need `metric` added to its signature, but that's all...?

Reply:
What do you mean? I am afraid that I am not following

Reply:
The function `plot_metric_over_time` will re-declare the inner function `get_model_group_evaluations` every time that the former is invoked. This is sometimes useful, but problems with it include:

* performance: the interpreter must recreate the function dynamically
* debugging & legibility: the inner function inherits a dynamic scope from every invocation of `plot_metric_over_time`
* testing & reusability: (if necessary), the inner function couldn't be either tested or reused with this factoring

Though the inner function here isn't huge, I would say it's big enough that it may as well be moved to the module scope, to avoid the above pitfalls.

Reply:
Makes sense! I will do it. 

My idea was not to pollute the module namespace with this function that is only used inside this one, but, what are you saying makes a lot of sense. What will be a good case for using  inner functions?

Reply:
Right! I think the same way. (But I've also been yelled at for doing so 😹)

The inner function can at least be given a leading underscore to indicate that it's not intended as part of the module API. 

I'd say the important thing is that inner functions be trivial – very short, (including lambdas of course).

But ultimately it's up to your judgement of the pros & cons relative to the situation.

Reply:
Thank you